geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java

```diff
@@ -86,12 +93,19 @@ public static int smove(RedisKey sourceKey, RedisKey destKey, byte[] member,
       RegionProvider regionProvider) {
     RedisSet source = regionProvider.getTypedRedisData(REDIS_SET, sourceKey, false);
     RedisSet destination = regionProvider.getTypedRedisData(REDIS_SET, destKey, false);
-    List<byte[]> memberList = new ArrayList<>();
-    memberList.add(member);
-    if (source.srem(memberList, regionProvider.getDataRegion(), sourceKey) == 0) {
+    if (!source.sismember(member)) {
       return 0;
     }
-    destination.sadd(memberList, regionProvider.getDataRegion(), destKey);
+    if (sourceKey.equals(destKey)) {
+      return 1;
+    }
+
+    List<byte[]> memberList = new ArrayList<>();
+    memberList.add(member);
+    RedisSet newSource = new RedisSet(source);
+    newSource.srem(memberList, regionProvider.getDataRegion(), sourceKey);
+    RedisSet newDestination = new RedisSet(destination);
+    newDestination.sadd(memberList, regionProvider.getDataRegion(), destKey);
```

Comment:
This method can be improved to be more efficient and fix some incorrect behaviour similar to that in RPOPLPUSH:
```
    RedisSet source = regionProvider.getTypedRedisData(REDIS_SET, sourceKey, false);

    if (!source.sismember(member)) {
      return 0;
    }

    if (sourceKey.equals(destKey)) {
      return 1;
    }

    List<byte[]> memberList = new ArrayList<>();
    memberList.add(member);
    RedisSet newSource = new RedisSet(source);
    newSource.srem(memberList, regionProvider.getDataRegion(), sourceKey

    RedisSet destination = regionProvider.getTypedRedisData(REDIS_SET, destKey, false);
    RedisSet newDestination = new RedisSet(destination);
    newDestination.sadd(memberList, regionProvider.getDataRegion(), destKey);
    return 1;
```

Without these changes, the below integration test fails with geode-for-redis but passes for open source Redis:
```
  @Test
  public void smove_withNonExistentSourceAndWrongTypeDestination_returnsZero() {
    jedis.set(DESTINATION_KEY, "not a RedisSet");

    assertThat(jedis.smove(NON_EXISTENT_SET_KEY, DESTINATION_KEY, MOVED_MEMBER))
        .isEqualTo(0);
  }
```

Reply:
@Donal... I am okay with the improvement but the only part I previously had issues with was the late instantiation of the `destination ` field. I had a TCL test fail on that as `getTypedRedisData(REDIS_SET, destKey, false)` had some checks on the Redisdata type. When we have that check before we even try to check for `sourceKey.equals(destKey)` we get the correct error of having a wrong type of data.

Reply:
Also, I think this test has two checks that it's doing but the wrong type error on destination takes precedence and it might not be required as we have other tests to do that.

Reply:
> Also, I think this test has two checks that it's doing but the wrong type error on destination takes precedence and it might not be required as we have other tests to do that.

Nonexistent key isn't an error condition, since it should just return 0 rather than an error message, so the test is only checking that we don't return the WRONGTYPE error incorrectly, not that we return one of two possible errors.

I see why the TCL test failed with my changes though; it's because we'e checking if the source set contains the member and returning early if it doesn't rather than checking if the source is null and then checking if it contains the member.

Right now we have one check with two possibilities for early return: the source set exists but doesn't contain the member, or the source set doesn't exist at all. In the second case, we should always return 0, regardless of what type the destination set is, as that's what open source Redis does. In the first case, we should check if the destination is the correct type before checking if the member is contained in the source set. The below code and additional integration test cover all the behaviour:
```
    RedisSet source = regionProvider.getTypedRedisData(REDIS_SET, sourceKey, false);

    if (source.isNull()) {
      return 0;
    }

    if (sourceKey.equals(destKey)) {
      return 1;
    }

    RedisSet destination = regionProvider.getTypedRedisData(REDIS_SET, destKey, false);
    if (!source.sismember(member)) {
      return 0;
    }
    List<byte[]> memberList = new ArrayList<>();
    memberList.add(member);
    RedisSet newSource = new RedisSet(source);
    newSource.srem(memberList, regionProvider.getDataRegion(), sourceKey);

    RedisSet newDestination = new RedisSet(destination);
    newDestination.sadd(memberList, regionProvider.getDataRegion(), destKey);
    return 1;
```

```
  @Test
  public void smove_withNonExistentMemberInSourceAndDestinationNotASet_returnsWrongTypeError() {
    String nonExistentMember = "foo";
    jedis.sadd(SOURCE_KEY, SOURCE_MEMBERS);
    jedis.set(DESTINATION_KEY, "not a set");

    assertThat(jedis.smove(NON_EXISTENT_SET_KEY, DESTINATION_KEY, nonExistentMember))
        .isEqualTo(0);
    assertThatThrownBy(() -> jedis.sismember(DESTINATION_KEY, nonExistentMember)).hasMessage(ERROR_WRONG_TYPE);
  }
```